src/ripple/shamap/README.md

```diff
@@ -26,164 +56,292 @@ There are two different ways of building and using a SHAMap:
-entirely evaporates when it is destroyed).  Nodes, once introduced to the
-immutable SHAMap, also never change their location in memory.  So nodes in
-an immutable SHAMap can be handled using raw pointers (if you're careful).
+So, somewhat counter-intuitively, an immutable SHAMap may grow as new nodes are
+introduced.  But an immutable SHAMap will never get smaller (until it entirely
+evaporates when it is destroyed).  Nodes, once introduced to the immutable
+SHAMap, also never change their location in memory.  So nodes in an immutable
+SHAMap can be handled using raw pointers (if you're careful).
 
 One consequence of this design is that an immutable SHAMap can never be
-"trimmed".  There is no way to identify unnecessary nodes in an immutable
-SHAMap that could be removed.  Once a node has been brought into the
-in-memory SHAMap, that node stays in memory for the life of the SHAMap.
+"trimmed".  There is no way to identify unnecessary nodes in an immutable SHAMap
+that could be removed.  Once a node has been brought into the in-memory SHAMap,
+that node stays in memory for the life of the SHAMap.
 
-Most SHAMaps are immutable, in the sense that they don't modify or remove
-their contained nodes.
+Most SHAMaps are immutable, in the sense that they don't modify or remove their
+contained nodes.
 
 An example where a mutable SHAMap is required is when we want to apply
-transactions to the last closed ledger.  To do so we'd make a mutable
-snapshot of the state tree and then start applying transactions to it.
-Because the snapshot is mutable, changes to nodes in the snapshot will not
-affect nodes in other SHAMAps.
+transactions to the last closed ledger.  To do so we'd make a mutable snapshot
+of the state trie and then start applying transactions to it. Because the
+snapshot is mutable, changes to nodes in the snapshot will not affect nodes in
+other SHAMaps.
+
+An example using a immutable ledger would be when there's an open ledger and
+some piece of code wishes to query the state of the ledger.  In this case we
+don't wish to change the state of the SHAMap, so we'd use an immutable snapshot.
+
+## Sequence numbers ##
+
+Both SHAMap's and their nodes carry a sequence number.  This is simply an
+unsigned number that indicates ownership or membership, or a non-membership.
 
-An example using a immutable ledger would be when there's an open ledger
-and some piece of code wishes to query the state of the ledger.  In this
-case we don't wish to change the state of the SHAMap, so we'd use an
-immutable snapshot.
+SHAMap's sequence numbers normally start out as 1.  However when a snap-shot of
+a SHAMap is made, the copy's sequence number is 1 greater than the original.
 
+The nodes of a SHAMap have their own copy of a sequence number.  If the SHAMap
+is not immutable, meaning it can change, then all of its nodes must have the
+same sequence number as the SHAMap itself.  This enforces an invariant that none
```

Comment:
`SHAMap`s don't have parents and siblings.  The node sequence number is more or less just a sharable flag.  0 means sharable.  Equal to the SHAMap's sequence number means not sharable.

Reply:
There is no explicit parent, which is why I put it in quotes, but the majority of sites constructing a `SHAMap` use `snapShot`, where I consider `this` to be an implicit "parent", instead of a constructor. Is `snapShot` the only way that two `SHAMap`s can share a node? There doesn't seem to be any method to add a `SHAMapAbstractNode` constructed from outside the `SHAMap`, so there would be no way to pass a node from one `SHAMap` to another; the only way to share a node is for the second `SHAMap` to inherit it from the first because it was constructed via `second = first.snapShot(isMutable)`.